### PR TITLE
Add haste support for log command - Update 2.1.16

### PR DESCRIPTION
**Pull request by TheMeinerLP**

### CODE REVIEW

**Comment by TheMeinerLP** [comment]
You can add as self in the Config a Haste Server

**Comment by juliarn** [comment]
I know, but I mean as a default

**Comment by TheMeinerLP** [comment]
Maybe optional update or pr

**Comment by GiantTreeLP** [comment]
This needs to be rebased and the unnecessary commits need to be removed.

**Comment by GiantTreeLP** [comment]
As with the other pull requests: please remove unnecessary files and commits that are part of other pull requests.

**Comment by TheMeinerLP** [comment]
How i can do it this ?

**Comment by GiantTreeLP** [comment]
Rebase only the relevant changes onto the development branch and skip any unnecessary commits.
This might require some manual patching.

**Comment by TheMeinerLP** [comment]
Fixed, and removed unnecessary commits.

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java [hunk 1]
```diff
@@ -124,9 +132,6 @@ private void defaultInit(ConsoleReader consoleReader) throws Exception
         configuration.set("server.webservice.hostaddress", hostName);
         configuration.set("server.webservice.port", 1420);
 
-        configuration.set("cloudnet-statistics.enabled", true);
-        configuration.set("cloudnet-statistics.uuid", UUID.randomUUID().toString());
-
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java:203]: Why did you remove this?

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 2]
```diff
@@ -4,49 +4,138 @@
...
+    switch (args.length) {
+      case 1: {
+        MinecraftServer minecraftServer = CloudNet.getInstance().getServer(args[0]);
+        if (minecraftServer != null) {
+          String rndm = NetworkUtils.randomString(10);
+          CloudNet.getInstance().getServerLogManager().append(rndm, minecraftServer.getServerId());
+          List<String> hasteServer = CloudNet.getInstance().getConfig().getHasteServer();
+          Random r = new Random();
+          sendPaste(sender, rndm, hasteServer.get(r.nextInt(hasteServer.size())), minecraftServer);
+          String x = new StringBuilder("http://")
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: I think a round-robin approach is more fitting here.
Try each server until a request succeeds (which is usually the first one)

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 3]
```diff
@@ -4,49 +4,138 @@
...
+   * Send the log to a random past site
+   * @param sender The sender of the command
+   * @param random A random string
+   * @param url The url of the paste site
+   * @param minecraftServer The minecraft server with the log inside
+   */
+  private void sendPaste(CommandSender sender, String random, String url,
+      MinecraftServer minecraftServer) {
+    final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    executorService.schedule(() -> {
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: You can use the built-in scheduler for this.

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 4]
```diff
@@ -4,49 +4,138 @@
...
+      connection.setRequestProperty("Content-Type", "text/plain");
+      connection.setDoInput(true);
+      connection.setDoOutput(true);
+      connection.connect();
+      OutputStream outputStream = connection.getOutputStream();
+      outputStream.write(builder.toString().getBytes(StandardCharsets.UTF_8));
+      outputStream.flush();
+      outputStream.close();
+      InputStream inputStream = connection.getInputStream();
+      Gson g = new Gson();
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: You don't need to instantiate Gson every time this command is used.
Either use a constant or a global instance.

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java (hunk 1 above)
Comment (TheMeinerLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java:203]: Oh its from from Patch

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java (hunk 2 above)
Comment (TheMeinerLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: How it is the best way to do that ?

**Comment by TheMeinerLP** [comment]
Review pls
@GiantTreeLP

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java (hunk 2 above)
Comment (GiantTreeLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: You are still picking a random server.
Do this logic inside your asynchronous paste method and try each server until one succeeds.

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java (hunk 2 above)
Comment (TheMeinerLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: Look ich check all Server, after that i pick some random of the working Server!

**Comment by GiantTreeLP** [comment]
You need to rebase your PR on the latest changes.

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java [hunk 5]
```diff
@@ -22,6 +22,7 @@
 import de.dytanic.cloudnetcore.util.defaults.BungeeGroup;
 import de.dytanic.cloudnetcore.util.defaults.LobbyGroup;
 import jline.console.ConsoleReader;
+import lombok.Getter;
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java]: Please remove any traces of lombok

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java (hunk 5 above)
Comment (TheMeinerLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/CloudConfig.java]: Ok!

**Comment by TheMeinerLP** [comment]
Please look if the command ready for merge
@GiantTreeLP

**Comment by Sarsum** [comment]
Please add https://paste.dsyn.ga to the list, otherwise @realPanamo might be sad

**Comment by Sarsum** [comment]
Also, I think it would be great if you could either choose which PasteServer to use, or use a random one, e.g. with a Config

**Comment by GiantTreeLP** [comment]
> Also, I think it would be great if you could either choose which PasteServer to use, or use a random one, e.g. with a Config

It's already configurable. But I don't like choosing a random server.
I suggested a round-robin approach or fallback approach.
So you can sort the servers by preference.

**Comment by TheMeinerLP** [comment]
I can do a check on all url and use the order from the Config. Is this idea better @GiantTreeLP

**Comment by GiantTreeLP** [comment]
You still don't understand fallback or round-robin.

Try to upload to the first server.  
If this fails, use the next server.  
If that one fails, use the next server.  
This goes on, until you have either a successful upload or reached the end of the list.

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 6]
```diff
@@ -4,49 +4,160 @@
...
+  /**
+   * Push the log to url
+   * @param url The url to push it
+   * @param builder The content of the log
+   * @return The finish url
+   */
+  private String postTo(String url, StringBuilder builder) {
+    try {
+      URL pasteUrl = new URL(url);
+      HttpURLConnection connection = (HttpURLConnection) pasteUrl.openConnection();
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: Use try-with-resources here

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 7]
```diff
@@ -4,49 +4,160 @@
...
+      outputStream.close();
+      InputStream inputStream = connection.getInputStream();
+
+      JsonObject object = g.fromJson(new InputStreamReader(inputStream), JsonObject.class);
+      connection.disconnect();
+      return object.get("key").getAsString();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    return "KEY_NOT_FOUND";
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: Throw an exception and catch it properly.
Returning valid values or null is a code-smell.

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 8]
```diff
@@ -4,49 +4,160 @@
...
+  /**
+   * Check url is page available
+   * @param url The url to check
+   * @return Return true if exist and connectible, else return false
+   */
+  private boolean checkUrl(String url) {
+    HttpURLConnection.setFollowRedirects(false);
+    HttpURLConnection con = null;
+    try {
+      con = (HttpURLConnection) new URL(url).openConnection();
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: Use try-with-resources here. (Although, this method is obsolete)

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

**Comment by juliarn** [comment]
Hm, there are still some parts which use 2 spaces instead of 4. Would be nice if you could fix this.

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java (hunk 8 above)
Comment (TheMeinerLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: @GiantTreeLP nope, is not autocloseable

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java (hunk 6 above)
Comment (TheMeinerLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: @GiantTreeLP and same here, url/httpconnection is not autocloseable

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 9]
```diff
@@ -4,49 +4,160 @@
...
+      connection.setRequestMethod("POST");
+      connection.setRequestProperty("Content-Type", "text/plain");
+      connection.setDoInput(true);
+      connection.setDoOutput(true);
+      connection.connect();
+      OutputStream outputStream = connection.getOutputStream();
+      outputStream.write(builder.toString().getBytes(StandardCharsets.UTF_8));
+      outputStream.flush();
+      outputStream.close();
+      InputStream inputStream = connection.getInputStream();
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: Close the stream after using it.

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java (hunk 8 above)
Comment (GiantTreeLP) [reply cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: Can you close it in finally though?

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 10]
```diff
@@ -131,16 +131,16 @@ private boolean postTo(CommandSender sender, String url, StringBuilder builder)
...
 			outputStream.flush();
-			outputStream.close();
 			InputStream inputStream = connection.getInputStream();
-
 			JsonObject object = g.fromJson(new InputStreamReader(inputStream), JsonObject.class);
 			int responseCode = connection.getResponseCode();
 			if (responseCode == 200) {
 				connection.disconnect();
 				sender.sendMessage("You can see the log at: " + String.format(url + "/%s", object.get("key").getAsString()));
+				connection.disconnect();
```
Comment (GiantTreeLP) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java:224]: You already disconnected the connection two lines before this.

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

**Review by GiantTreeLP — APPROVED** [review]

File: cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java [hunk 11]
```diff
@@ -4,49 +4,259 @@
...
+			paste.append(String.format("User: %s", System.getProperty("user.name")))
+					.append('\n');
+			paste.append(String.format("OS: %s %s (%s)", System.getProperty("os.name"),
+					System.getProperty("os.arch"),
+					System.getProperty("os.version")))
+					.append('\n');
+			paste.append(String.format("Memory Used: %d Mb", (NetworkUtils.system().getTotalPhysicalMemorySize() -
+					NetworkUtils.system().getFreePhysicalMemorySize()) / 1024L)).append('\n');
+			paste.append(String.format("Memory Free: %d Mb", NetworkUtils.system().getFreePhysicalMemorySize() / 1024L)).append('\n');
+			paste.append(String.format("Memory Max: %d Mb", NetworkUtils.system().getTotalPhysicalMemorySize() / 1024L)).append('\n');
```
Comment (einrobin) [inline cloudnet-core/src/main/java/de/dytanic/cloudnetcore/command/CommandLog.java]: You're calculating the memory as KB, but displaying it as "Mb"

**Pull request merged**